cmd/snap-bootstrap/initramfs_install_handler.go

```diff
@@ -0,0 +1,418 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/asserts/sysdb"
+	"github.com/snapcore/snapd/boot"
+	"github.com/snapcore/snapd/gadget"
+	"github.com/snapcore/snapd/gadget/install"
+	"github.com/snapcore/snapd/interfaces/builtin"
+	"github.com/snapcore/snapd/kernel/fde"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+
+	// to set sysconfig.ApplyFilesystemOnlyDefaultsImpl
+	_ "github.com/snapcore/snapd/overlord/configstate/configcore"
+	"github.com/snapcore/snapd/overlord/devicestate"
+	"github.com/snapcore/snapd/progress"
+	"github.com/snapcore/snapd/secboot"
+	"github.com/snapcore/snapd/seed"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/snapfile"
+	"github.com/snapcore/snapd/systemd"
+	"github.com/snapcore/snapd/timings"
+)
+
+const (
+	rootdir          string = "/"
+	InitRamfsFdeHook string = "fde-reveal-key"
+)
+
+var (
+	// assertion store to be trusted
+	trusted = sysdb.Trusted()
+	// /run/mnt/ubuntu-data is not exported, why?
+	InstallUbuntuDataDir = filepath.Join(boot.InitramfsRunMntDir, "ubuntu-data")
+)
+
+// mounts info
+type mountInfo struct {
+	mountPoint string
+	unmountFnc func() error
+	// essential flag is tracking mounts essential for install mode
+	// those mounts should be destroyed when moving to run mode
+	essential bool
+}
+
+// tracking of active mounts
+var activeMounts []mountInfo
+
+func doUnmount(info *mountInfo) error {
+	if err := info.unmountFnc(); err != nil {
+		// only log error and try to unmount next
+		return fmt.Errorf("Failed to unmount%s, %v", info.mountPoint, err)
+	}
+	return nil
+}
+
+func hasFDESetupHook(kernelInfo *snap.Info) (bool, error) {
+	_, err := exec.LookPath(InitRamfsFdeHook)
+	return err == nil, err
+}
+
+func runFDESetupHook(req *fde.SetupRequest) ([]byte, error) {
+	stdin, err := json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf(`cannot build request %v for fde-reveal-key: %v`, req, err)
+	}
+	return fde.RunFDEinitramfsHelper(InitRamfsFdeHook, stdin)
```

Comment:
mmh, we are running the fde-reveal-key instead of the fde setup hook? We really need to think this through, this is not a viable change.

Reply:
anyway we have mounted the kernel so we have access to the other hook?